src/pieces.rs

```diff
@@ -1,112 +1,82 @@
-        match position {
-            Top => self.top = piece,
-            Middle => self.mid = piece,
-            Bottom => self.bottom = piece,
-            Empty => panic!("Cannot set a piece at TowerHeight::Empty")
-        }
-
-        if !self.is_valid() {
-            panic!("Attempt to set Tower to an invalid state: {:?}", self);
-        }
-    }
-
-    /// Removes and returns the top most piece from the tower
-    /// Panics if the tower is empty
-    pub fn pop(&mut self) -> Piece<'a> {
-        let height = self.height();
-
-        if height == TowerHeight::Empty {
-            panic!("Cannot pop an empty tower!")
-        }
-        // This unwrap is safe because the tower is non-empty
-        let top_piece = self.get(height).unwrap();
-        self.set(None, height);
-        return top_piece
-    }
-
-    /// Adds a piece to the top most position on the tower
-    /// Panics if the tower is full
-    pub fn drop_piece(&mut self, piece: Piece<'a>) {
-        let height = self.height();
-        use pieces::TowerHeight::*;
-        match height {
-            Top => panic!("Tower is full."),
-            Middle => self.set(Some(piece), Top),
-            Bottom => self.set(Some(piece), Middle),
-            Empty =>self.set(Some(piece), Bottom),
+    /// Returns a tower that has a piece added to the topmost position on this tower
+    /// Returns Err if the tower is full does not modify Tower state when this happens
+    /// This function does not modify the original tower.
+    pub fn drop_piece(&self, piece: Piece<'a>) -> Result<Tower<'a>, &'static str> {
+        use pieces::Tower::*;
+        match *self {
+            Empty => Ok(Single(piece)),
+            Single(bottom) => Ok(Double(bottom, piece)),
+            Double(bottom, middle) => Ok(Triple(bottom, middle, piece)),
+            Triple(_, _, _) => Err("Tower is full."),
         }
     }
 
     pub fn height(&self) -> TowerHeight {
-        if let Some(_) = self.top {
-            return TowerHeight::Top;
-        } else if let Some(_) = self.mid {
-            return TowerHeight::Middle;
-        } else if let Some(_) = self.bottom {
-            return TowerHeight::Bottom;
-        } else {
-            return TowerHeight::Empty;
+        use pieces::Tower::*;
+        match *self {
+            Empty => TowerHeight::Empty,
+            Single(_) => TowerHeight::Bottom,
+            Double(_, _) => TowerHeight::Middle,
+            Triple(_, _, _) => TowerHeight::Top
         }
     }
 
     /// A tower is valid as long as no two pieces from the same player
     /// of the same type are in it
-    ///    For example, (Your) Pawn, (Your) Gold, (Your) Gold is disallowed
-    ///    but (Your) Pawn, (Your) Gold, (Enemy) Gold is fine
+    /// For example, (Your) Pawn, (Your) Pawn is disallowed
+    /// but (Your) Pawn, (Enemy) Pawn is fine
+    /// ```
+    /// let player1 = Player::new_blank();
+    /// let player2 = Player::new_blank();
```

Comment:
Just a note, I don't know if you necessarily want to do it in this case, but you can hide lines in the doctests by starting them with `#`, so if you decided that it's obvious that the players were `Player` objects, you could hide their construction.